### DISCUSSION TITLE
API Timeouts when submitting transactions

### DISCUSSION THREAD

**Discussion by prononext** [firefly-iii/firefly-iii#7719] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I am trying to use the Firefly III API to get about 2000 transactions from a long time period into my system and I am failing for 2 days in a row. 

Using version-6.0.16 and running as docker with mariadb. I am all on standard settings and have also activated the debug log in the environment variables. 

On my api calls I got to the point to split the calls in batches and write 1 transaction per second but I never get over 800 to POST.

 ` connect ETIMEDOUT 192.168.1.10:8080`

I also tried another approch then to compare the posted transactions to the new ones and only post the missing ones, even like this I do not even get to the point that my GET request can call all pages, even with a 30 second wait after each page.

 ` connect ETIMEDOUT 192.168.1.10:8080`

I also have tried setting the timeout inside my api call to 60 seconds which did not change anything. 
Also tried to make different sets of 2000 split in each 100 but then after a couple of houndret posted.... same error. 
As I also cannot compare create while the GET for several pages also result in a timeout I am quite helpless here. 
I buped my vm up to 8 cores with 8GB ram but nothing changed. Memory usage is under 3GB and processor usage under 30% in total. On other hosts I do api calls with 10thousands of posts and gets without any problems.

I have the feeling that there is some kind of limiter active in the webserver of the image?

Does anybody have a clue to how to get rid of this timeouts?

Thank you very much. 

### Debug information

the logs are just normal with no errors at all showing the post request coming in and then just stop. even the container is staying online and the GUI is available.

### Expected behaviour

POST and GET as many as you like without hitting any timeout as long as your hardware is able to serve the requests. 

### Steps to reproduce

Spinup a docker stack with the standard compose and env.
Make at least a couple of houndret POST and GET requests to the api. 
Probably hitting the timeout.

**Comment by JC5** [comment]
Ok first, let's use a title that helps other users when searching for information as well.

**Comment by JC5** [comment]
>  Make at least a couple of houndret POST and GET requests to the api.

I'm currently 2000 transactions in (5 threads) and so far nothing. Speed per transaction is stable at around a quarter of a second. The UI is slower, the index takes 3 seconds to load.

**Comment by JC5** [comment]
Here's my script. Still running smoothly.

```php
<?php

require 'vendor/autoload.php';

function randomString(int $int): string
{
    return substr(md5(rand()), 0, $int);
}

function randomAmount(int $low, int $high)
{
    $low = $low * 100;
    $high = $high * 100;
    return random_int($low, $high) / 100;
}
$count = 0;
while (true) {
    $start = microtime(true);
    $transaction = [
        'transactions' => [
            [
                'type' => 'withdrawal',
                'source_id' => 1,
                'date' => '2023-07-04',
                'destination_name' => randomString(10),
                'description' => randomString(20),
                'amount' => randomAmount(1, 1000),
                'category_name' => randomString(10),
            ],
        ],
    ];

// post to Firefly III
    $token = trim(file_get_contents('token.txt'));
    $url = 'http://localhost/api/v1/transactions';
    $opts = [
        'json' => $transaction,
        'headers' => [
            'Accept' => 'application/json',
            'Content-Type' => 'application/json',
            'Authorization' => sprintf('Bearer %s', $token),
        ],
    ];


    $client = new \GuzzleHttp\Client();
    $client->post($url, $opts);
    $time = round(microtime(true) - $start, 2);

    $count++;
    echo sprintf('%d: Posted transaction in %s seconds', $count, $time);
    echo PHP_EOL;
}

```

**Comment by prononext** [comment]
ok I will give it a try. I also will do a complete reset of the install. Using n8n to do the POSTs.

**Comment by prononext** [comment]
@JC5 Just spin up a new instance and tried again. Got untill 931 transactions untill the timeout. 

Thats the log for the last transaction.

`[2023-07-04T20:07:35.141876+02:00] production.INFO: AUDIT: Store new transaction over API.` with no error

After that just a normal health check. And I am not using any proxy or balancer in front of firefly.
`10.0.0.2 - - [04/Jul/2023:20:07:35 +0200] "POST /api/v1/transactions HTTP/1.1" 200 2770 "-" "axios/0.21.4"
127.0.0.1 - - [04/Jul/2023:20:07:53 +0200] "GET /health HTTP/1.1" 200 1776 "-" "Firefly III Health Checker/2.0"`

**Reply by JC5** [reply]
Sorry, I don't really have an idea either.

**Comment by prononext** [comment]
I FINALLY got it working!

**I put a Traefik and HAProxy stack in front of the Firefly instance and all POST and GET requests worked without any error.** 

I cannot explain that, but yeah sometimes its like this....

Thanks @JC5 for this beautiful Software. Hope that v3 GUI is out soon 💯 👍
